scipy/signal/_signaltools.py

```diff
@@ -3559,20 +3562,28 @@ def detrend(data, axis=-1, type='linear', bp=0, overwrite_data=False):
     0.06  # random
 
     """
+    xp = array_namespace(data)
     if type not in ['linear', 'l', 'constant', 'c']:
         raise ValueError("Trend type must be 'linear' or 'constant'.")
-    data = np.asarray(data)
-    dtype = data.dtype.char
-    if dtype not in 'dfDF':
-        dtype = 'd'
+    data = xp.asarray(data)
+    dtype = data.dtype
+    if data.dtype not in [xp.float32, xp.float64, xp.complex128, xp.complex64]:
+        dtype = xp.float64
     if type in ['constant', 'c']:
-        ret = data - np.mean(data, axis, keepdims=True)
+        ret = data - xp.mean(xp.asarray(data, dtype=dtype), axis=axis, keepdims=True)
         return ret
     else:
         dshape = data.shape
         N = dshape[axis]
-        bp = np.sort(np.unique(np.concatenate(np.atleast_1d(0, bp, N))))
-        if np.any(bp > N):
+        if isinstance(bp, int):
+            bp = xp.sort(xp.unique(xp.asarray([0, bp, N])))
+        else:
+            bp = xp.asarray(bp)
+            new_bp = xp.empty(size(bp) + 1)
+            new_bp[:size(bp)] = bp
+            new_bp[size(bp)] = N
+            bp = xp.sort(xp.unique(xp.asarray([0] + new_bp)))
```

Comment:
Isn't there something that we can avoid doing this type of back and forth? 

I found this https://data-apis.org/array-api/latest/API_specification/generated/array_api.unique_values.html?highlight=unique#array_api.unique_values but it doesn't have the NumPy functionality. 

For example, `numpy.unique` has `return_index` keyword to get rid of the sorting need but I don't know if array API supports it which is my biggest concern adopting this api; potential code divergence later in different packages. Mini demo;

```python
>>> import numpy as np
>>> arr = np.array([1.0, np.sqrt(2), np.sqrt(2), 1/3, 1/3, np.pi])
>>> arrun, ind = np.unique(arr, return_index=True)
>>> arr[ind]
array([0.33333333, 1.        , 1.41421356, 3.14159265])
```

So indeed you can embed `bp` into a larger array

```python
new_bp = xp.empty(size(bp)+2, dtype=dtype)
new_bp[0] = 0
new_bp[-1] = N  # -1 is not working anymore in array_api?
new_bp[1:-1] = bp 
```

I know `bp` is typically a small array but still the code gets really complicated for a relatively simple operation.